src/lib/moment/start-end-of.js

```diff
@@ -1,59 +1,127 @@
+var MS_PER_400_YEARS = (365 * 400 + 97) * MS_PER_24_HOUR_DAY;
+
+// actual modulo - handles negative numbers (for dates before 1970):
+function mod(dividend, divisor) {
+    return (dividend % divisor + divisor) % divisor;
+}
+
+function localStartOfDate(y, m, d) {
+    // the date constructor remaps years 0-99 to 1900-1999
+    if (0 <= y && y <= 99) {
+        return new Date(y + 400, m, d) - MS_PER_400_YEARS;
+    } else {
+        return new Date(y, m, d).valueOf();
+    }
+}
+
+function utcStartOfDate(y, m, d) {
+    // Date.UTC remaps years 0-99 to 1900-1999
+    if (0 <= y && y <= 99) {
+        return Date.UTC(y + 400, m, d) - MS_PER_400_YEARS;
+    } else {
+        return Date.UTC(y, m, d);
+    }
+}
 
 export function startOf (units) {
+    var time;
     units = normalizeUnits(units);
-    // the following switch intentionally omits break keywords
-    // to utilize falling through the cases.
+    if (units === undefined || units === 'millisecond' || !this.isValid()) {
+        return this;
+    }
+
+    var startOfDate = this._isUTC ? utcStartOfDate : localStartOfDate;
+
     switch (units) {
         case 'year':
-            this.month(0);
-            /* falls through */
+            time = startOfDate(this.year(), 0, 1);
+            break;
         case 'quarter':
+            time = startOfDate(this.year(), this.month() - this.month() % 3, 1);
+            break;
         case 'month':
-            this.date(1);
-            /* falls through */
+            time = startOfDate(this.year(), this.month(), 1);
+            break;
         case 'week':
+            time = startOfDate(this.year(), this.month(), this.date() - this.weekday());
+            break;
         case 'isoWeek':
+            time = startOfDate(this.year(), this.month(), this.date() - (this.isoWeekday() - 1));
+            break;
         case 'day':
         case 'date':
-            this.hours(0);
-            /* falls through */
+            time = startOfDate(this.year(), this.month(), this.date());
+            break;
         case 'hour':
-            this.minutes(0);
-            /* falls through */
+            time = this._d.valueOf();
+            time = (time - mod(time + (this._isUTC ? 0 : this.utcOffset() * MS_PER_MINUTE), MS_PER_HOUR));
+            break;
         case 'minute':
-            this.seconds(0);
-            /* falls through */
+            time = this._d.valueOf();
+            this._d.setTime(time - mod(time, MS_PER_MINUTE));
+            return this;
         case 'second':
-            this.milliseconds(0);
-    }
-
-    // weeks are a special case
-    if (units === 'week') {
-        this.weekday(0);
-    }
-    if (units === 'isoWeek') {
-        this.isoWeekday(1);
-    }
-
-    // quarters are also special
-    if (units === 'quarter') {
-        this.month(Math.floor(this.month() / 3) * 3);
+            time = this._d.valueOf();
+            this._d.setTime(time - mod(time, MS_PER_SECOND));
```

Comment:
same here.

In general most of the switch uses break, and only the last 2 use return, which is confusing. I'm sure the perf will be fine.